src/s_inter.c

```diff
@@ -201,30 +740,70 @@ static int sys_domicrosleep(int microsec, int pollem)
 {
     struct timeval timout;
     int i, didsomething = 0;
-    t_fdpoll *fp;
     timout.tv_sec = 0;
     timout.tv_usec = 0;
     if (pollem && pd_this->pd_inter->i_nfdpoll)
     {
-        fd_set readset, writeset, exceptset;
-        FD_ZERO(&writeset);
-        FD_ZERO(&readset);
-        FD_ZERO(&exceptset);
-        for (fp = pd_this->pd_inter->i_fdpoll,
-            i = pd_this->pd_inter->i_nfdpoll; i--; fp++)
-                FD_SET(fp->fdp_fd, &readset);
-        if(select(pd_this->pd_inter->i_maxfd+1,
-                  &readset, &writeset, &exceptset, &timout) < 0)
-          perror("microsleep select");
+#ifdef THREADED_IO
+        // Pd was built with THREADED_IO support, but no one has taken
+        // charge of the IO duties, so we do it ourselves in the audio thread.
+        // Sub-optimal because of the extra memory copies, but better than
+        // nothing.
+        if(!pd_this->pd_inter->i_dontmanageio)
+            sys_doio(pd_this);
+        ring_buffer* rbrmfdsend = pd_this->pd_inter->i_rbrmfdsend;
+        if(rbrmfdsend)
+        {
+            // fds queued for deletion upon a failed send() call
+            while(rb_available_to_read(rbrmfdsend) >= sizeof(int))
+            {
+                int fd;
+                rb_read_from_buffer(rbrmfdsend, (char*)&fd, sizeof(fd));
+                printf("Removing send fd %d from send list\n", fd);
+                // find what object owns it
+                for(i = 0; i < pd_this->pd_inter->i_nfdsend; ++i)
+                {
+                    t_fdsend* fdsends = pd_this->pd_inter->i_fdsend;
+                    if(fd == fdsends[i].fds_fd)
+                    {
+                        printf("fd %d found to belong to %p, so calling %p\n", fd, fdsends[i].fds_ptr, fdsends[i].fds_fn);
+                        fdsends[i].fds_fn(fdsends[i].fds_ptr);
+                    }
+                }
+            }
+        }
+#endif // THREADED_IO
+        fd_set readset;
+        fdp_select(&readset, timout, kFdpManagerAudioThread);
```

Comment:
What's the reason for calling `fdp_select` here? We already call `fdp_select` in the I/O thread and set `fdp_manager` to `kFdpManagerAudioThread` if the socket is available for reading. Why do we need to call `select` again?

Reply:
I guess this is only needed if Pd is compiled without `THREADED_IO`?

Reply:
This was a failsafe to ensure that the data is actually there in the socket(see my message above about the `pollfn()` failing when called without data). I figured it wouldn't be much overhead, as it would only happen sporadically and only when at least one `fdpoll` has been marked as ready by the other thread, so the real-time guarantees will be lost in the `pollfn()` anyhow.

Sorting out the possible threading issues with atomics above would probably make this redundant when `THREADED_IO` is enabled.

Reply:
> Sorting out the possible threading issues with atomics above would probably make this redundant when THREADED_IO is enabled.

I think so too.

Reply:
https://github.com/giuliomoro/pure-data-1/commit/e38a0db192667ca4bc7fa87400641003614b08b1